### PR TITLE
MaxNumberOfParallelism has been added

### PR DESCRIPTION
**Pull request by AlexeyBogdan95**

### CODE REVIEW

File: src/SqsPoller/SqsPollerHostedService.cs [hunk 1]
```diff
@@ -62,6 +63,7 @@ private async Task Handle(string queueUrl, CancellationToken cancellationToken)
                 _logger.LogTrace(
                     "Start processing the message with id {message_id} and ReceiptHandle {receipt_handle}");
 
+                await semaphore.WaitAsync(cancellationToken);
                 await _consumerResolver
```
Comment (IhnatKlimchuk) [inline src/SqsPoller/SqsPollerHostedService.cs]: Here you create a chain of tasks by using `ContinueWith` that works like: A -> B(A) -> C(B)-> D(C) -> E(D). As result you `await` task E. That probably will lead to one by one scheduling. Correct me If I'm wrong.

File: src/SqsPoller/SqsPollerHostedService.cs [hunk 2]
```diff
@@ -78,6 +80,9 @@ await _consumerResolver
                         async _ => await DeleteMessage(queueUrl, message, cancellationToken),
```
Comment (IhnatKlimchuk) [inline src/SqsPoller/SqsPollerHostedService.cs]: Probably here `async/await` can lead to not catching exceptions. Example: https://stackoverflow.com/questions/18697167/use-an-async-callback-with-task-continuewith

File: src/SqsPoller/SqsPollerHostedService.cs (hunk 1 above)
Comment (dsharkou) [reply src/SqsPoller/SqsPollerHostedService.cs]: Also with async/await we lost simultaneous handling of events. So in that case I think semaphore is useless

File: src/SqsPoller/SqsPollerConfig.cs [hunk 3]
```diff
@@ -62,5 +62,10 @@ public class SqsPollerConfig
         /// </summary>
         /// <inheritdoc cref="https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-attributes.html"/>
         public List<string> MessageAttributeNames { get; set; } = new List<string> {"All", ".*"};
+
+        /// <summary>
+        /// Gets and sets MaxNumberOfMessages (default 100).
```
Comment (dsharkou) [inline src/SqsPoller/SqsPollerConfig.cs]: Looks like typo - should be `MaxNumberOfParallelism`

File: src/SqsPoller/SqsPollerConfig.cs [hunk 4]
```diff
@@ -62,5 +62,10 @@ public class SqsPollerConfig
...
+        /// </summary>
+        public int MaxNumberOfParallelism { get; set; } = 100;
```
Comment (dsharkou) [inline src/SqsPoller/SqsPollerConfig.cs:69]: I see it's not used anywhere

File: src/SqsPoller/SqsPollerHostedService.cs [hunk 5]
```diff
@@ -62,22 +64,31 @@ private async Task Handle(string queueUrl, CancellationToken cancellationToken)
...
+                                    "Failed to handle message with id {message_id} and ReceiptHandle {receipt_handle}");
+                            }
+                            else if (task.IsCanceled)
+                            {
+                                _logger.LogWarning(
+                                    "Failed to handle message with id {message_id} and ReceiptHandle {receipt_handle}. Task has been cancelled");
+                            }
+                            else
+                            {
+                                DeleteMessage(queueUrl, message, cancellationToken);
```
Comment (dsharkou) [inline src/SqsPoller/SqsPollerHostedService.cs]: I think we should avoid using cancellation token because in case of message was handled successfully

**Review by dsharkou — CHANGES_REQUESTED** [review]

**Review by dsharkou — APPROVED** [review]

File: src/SqsPoller/SqsPollerHostedService.cs [hunk 6]
```diff
@@ -62,22 +63,31 @@ private async Task Handle(string queueUrl, CancellationToken cancellationToken)
...
+                            {
+                                _logger.LogWarning(
+                                    "Failed to handle message with id {message_id} and ReceiptHandle {receipt_handle}. Task has been cancelled");
+                            }
+                            else
+                            {
+                                DeleteMessage(queueUrl, message);
+                            }
+
+                            semaphore.Release();
```
Comment (IhnatKlimchuk) [inline src/SqsPoller/SqsPollerHostedService.cs:87]: Does it make sense put `semaphore.Release();` in `finally` so if `DeleteMessage` throws nothing is lost?

**Review by IhnatKlimchuk — APPROVED** [review]

**Pull request merged**